File: Data_Structures/classes.js
// What are data structures?
// Values, relationships between them, and funcs that can be applied to the data

class Student {
  constructor(firstName, lastName, year) {
    this.firstName = firstName;
    this.lastName = lastName;
    this.year = year;
    this.late = 0;
    this.scores = [];
  }

  // These are called instance methods
  fullName() {
    return `Your first name is ${this.firstName} ${this.lastName}`;
  }

  markLate() {
    this.late += 1;
    if (this.late >= 3) return "You're expelled!!!";
    return `${this.firstName} ${this.lastName} has been late ${this.late}`;
  }

  addScore(score) {
    this.scores.push(score);
    return this.scores;
  }

  calculateAverage() {
    let sum = this.scores.reduce((a, b) => a + b);
    return sum / this.scores.length;
  }

  // Can also define class methods, basically a utility function that doesn't
  // Depend on a specific instance

  static EnrollStudents() {
    return "ENROLLING STUDENTS FAM";
  }
}

let firstStudent = new Student("Colt", "Steele", 4);

firstStudent.addScore(68);
firstStudent.addScore(85);
console.log(firstStudent.addScore(88));
console.log(firstStudent.calculateAverage());

console.log(Student.EnrollStudents());

File: Problem_Solving/classes.py
class Student: 
    def __init__(self, firstName, lastName, year):
        self.firstName = firstName
        self.lastName = lastName
        self.year = year
        self.late = 0
    
    def fullName(self):
        return self.firstName + " " + self.lastName
    
    def addLates(self):
        self.late+=1
        return 'You have been late {lates} times'.format(lates = self.late)

firstStudent = Student("Hassan", "El Mghari", 5)
firstStudent.addLates()
firstStudent.addLates()
firstStudent.addLates()
firstStudent.addLates()
print(firstStudent.addLates())